### ISSUE TITLE
Add LogVal to MCState 

### ISSUE THREAD

**Issue by gcarleo** [netket/netket#612]
In order to have a complete interface in ```MCState``` (and also to easily implement some missing features in 3.0, such as #556 ) it is important to add a (differentiable) method to ```MCState``` that return the logarithm of the variational state in a given basis. 

I would like to discuss here about the signature of this. I had in mind something like 

```python
def LogVal(x,R=None):
      """Returns Log <x|R|Psi>, where R is an optional rotation operator""" 
```

Alternatively, a more general interface would be something like that 

```python
def LogInnerProduct(phi):
      """Returns Log <Phi|Psi>"""
```

however I guess the issue is implementing the latter to be aware of cases where Phi is just a state of the computational basis and (or a linear combination of a small number of those) and not a generic ```VariationalState```

**Comment by femtobit** [comment]
> however I guess the issue is implementing the latter to be aware of cases where Phi is just a state of the computational basis

I think for this reason it makes sense to have two separate methods - a simple one that just computes the log-amplitude and a more general log-inner-product function for cases where that is not enough.

As far as names go, maybe we could be a bit more precise than "log val" and have something like
```python
def log_amplitude(x, *, R=None):
    """Evaluate the variational wave function at configuration `x`, giving 〈x|R|ψ〉."""
    ...

def log_overlap(phi):  # or log_inner_product
    """Estimate the overlap 〈ɸ|ψ〉."""
    ...
```

**Comment by gcarleo** [comment]
yes I agree... the only issue is that in principle it would still make sense to do ```log_overlap(x)``` ... :) x is just a very special ```VariationalState```. Maybe (not sure) this is telling us that we should extend a bit the hierarchy of states to have something like

```QuantumState->BasisState```
```QuantumState->VariationalState...``` etc

**Comment by PhilipVinc** [comment]
--
Right now `MCState` has a method called `MCState.evaluate(x)` which is essentially `logVal`.
Indeed if the method was called `log_val` it would be more clear I guess?

It's not used anywhere, as it's only a user-facing api, so changing it's name is not important, but I would like to settle on the name before the school of Thursday and not change it in the future.

In the future we might add the rotation kwarg if needed.

Can we just settle on a name for it? I like `log_amplitude`, personally. We could also go with `log_value` or `log_val` or `logval`.

The standard in netket right now is to use the `this_thing_convention` rather than camelaase so `logval` should be excluded.
I'd also exclude `log_val` over `log_value` because in general we don't use abbreviations.

So I guess the fight is between `log_value` vs `log_amplitude`.

**Comment by femtobit** [comment]
> So I guess the fight is between `log_value` vs `log_amplitude`.

Agreed. I prefer `log_amplitude` to `log_value` (because "value" is very unspecific) but both would be fine.

**Comment by PhilipVinc** [comment]
Though, since in principle it would be nice to use the same method for Mixed states, or maybe POVM states if someone ever supports them, this won't be an amplitude anymore... yet the code will be the same.

So maybe `log_value` would be better.

**Comment by gcarleo** [comment]
indeed ```log_amplitude``` is specific for pure states ... but ```log_value``` is a bit too generic, I am not sure... let's do ```log_value``` for the moment and then we can just add ```log_amplitude``` as a convenience method in case (definitely ```evaluate``` is not a good name instead)

**Comment by PhilipVinc** [comment]
ok, i have renamed to `log_value` in #632

**Comment by PhilipVinc** [comment]
Re 
```python
def log_amplitude(x, *, R=None):
    """Evaluate the variational wave function at configuration `x`, giving 〈x|R|ψ〉."""
    ...
```

I was recently thinking about this.
I think a good enough API would be to use numpy broadcasting semantics and broadcast along the correct dimension of `x` and `R`.

So if x is a batch of states and R is one operator, we compute `out_i = <x_i|R|psi>` , while if R is also a vector with same leading dimension of x, then we get `out_i = <x_i|R_i|psi>`
